File: 백준/Silver/30804. 과일 탕후루/과일 탕후루.java
import java.io.*;
import java.util.*;

class Main{
    public static void main(String[] args)throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        int n = Integer.parseInt(br.readLine());
        int[] arr = new int[n];
        StringTokenizer st = new StringTokenizer(br.readLine());
        Map<Integer, Integer> map = new HashMap<>();
        for (int i = 0; i < n; i++) {
            int num = Integer.parseInt(st.nextToken());
            arr[i] = num;
        }

        int start = 0;
        int ans = 0;
        for(int end = 0; end < arr.length; end++) {
            map.put(arr[end], map.getOrDefault(arr[end], 0) + 1);

            while(map.size() > 2) {
                map.put(arr[start], map.get(arr[start]) - 1);
                if(map.get(arr[start]) == 0) map.remove(arr[start]);
                start++;
            }

            ans = Math.max(ans, end - start + 1);
        }
        System.out.println(ans);
    }
}

File: 백준/Silver/30804. 과일 탕후루/과일 탕후루.py
import sys
input = sys.stdin.readline

n = int(input())
arr = list(map(int, input().split()))
left = 0
ans = 0
dic = {}
for right in range(n):
    dic[arr[right]] = dic.get(arr[right], 0) + 1
    while len(dic) > 2:
        dic[arr[left]] -= 1
        if dic[arr[left]] == 0:
            del dic[arr[left]]
        left += 1
    ans = max(ans, right - left + 1)
print(ans)